Xamarin.Forms.Platform.Android/ImageCache.cs

```diff
@@ -0,0 +1,120 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Xamarin.Forms.Platform.Android
+{
+	/// <summary>
+	/// I setup the access to all the cache elements to be async because
+	/// if I didn't then it was locking up the GC and freezing the entire app
+	/// </summary>
+	class ImageCache
+	{
+		readonly FormsLruCache _lruCache;
+		readonly ConcurrentDictionary<string, SemaphoreSlim> _waiting;
+
+		public ImageCache() : base()
+		{
+			_waiting = new ConcurrentDictionary<string, SemaphoreSlim>();
+			_lruCache = new FormsLruCache();			
+		}
+
+		void Put(string key, TimeSpan cacheValidity, global::Android.Graphics.Bitmap cacheObject)
+		{
+			_lruCache.Put(key, new CacheEntry() { TimeToLive = DateTimeOffset.Now.Add(cacheValidity), Data = cacheObject });
+		}
+
+		public Task<Java.Lang.Object> GetAsync(string cacheKey, TimeSpan cacheValidity, Func<Task<Java.Lang.Object>> createMethod)
+		{
+			return Task.Run(async () =>
```

Comment:
Since you `await` the `SemaphoreSlim` can you drop this `Task.Run` now?

Reply:
My concern with that is if the winning thread (first one that loads image and puts it into the cache) holds the semaphore then it might not change threads to a Task Pool thread. Therefore, the code afterwards (mainly the Get) will run on the calling thread. Which I'd like to avoid.